packages/terminal-external/src/electron-browser/terminal-external-contribution.ts

```diff
@@ -0,0 +1,133 @@
+        });
+    }
+
+    registerKeybindings(keybindings: KeybindingRegistry): void {
+        keybindings.registerKeybinding({
+            command: TerminalExternalCommands.OPEN_NATIVE_CONSOLE.id,
+            keybinding: 'shift+ctrlcmd+c'
+        });
+    }
+
+    /**
+     * Open native console on host machine.
+     */
+    protected async openTerminalExternal(): Promise<void> {
+        const configuration = this.getTerminalExternalConfiguration();
+
+        // Multi-root workspace opened.
+        if (this.workspaceService.isMultiRootWorkspaceOpened) {
+            // Display a quick pick to let users choose which workspace to spawn the terminal.
+            const chosenWorkspaceRootUri = await this.selectTerminalExternalCwd();
+            if (chosenWorkspaceRootUri) {
+                await this.terminalExternalService.openTerminal(configuration, chosenWorkspaceRootUri.toString());
+            }
+            return;
+        }
+
+        // Only one workspace opened.
+        if (this.workspaceService.opened) {
+            // Spawn the terminal at the root of the current workspace.
+            const workspaceRootUri = this.workspaceService.tryGetRoots()[0].resource;
+            await this.terminalExternalService.openTerminal(configuration, workspaceRootUri.toString());
+            return;
+        }
+
+        // No workspaces opened.
+        const activeEditorUri = this.editorManager.activeEditor?.editor.uri;
+        if (activeEditorUri) {
+            // Spawn external terminal at the parent folder of active editor file.
+            await this.terminalExternalService.openTerminal(configuration, activeEditorUri.parent.toString());
+        } else {
+            // Spawn external terminal at user home directory if no workspaces opened and no current active editor.
+            const userHomeDir = await this.envVariablesServer.getHomeDirUri();
+            await this.terminalExternalService.openTerminal(configuration, userHomeDir);
+        }
+    }
+
+    /**
+     * Get the external terminal configurations from preferences.
+     */
+    protected getTerminalExternalConfiguration(): TerminalExternalConfiguration {
+        return {
+            'terminal.external.linuxExec': this.terminalExternalPreferences['terminal.external.linuxExec'],
+            'terminal.external.osxExec': this.terminalExternalPreferences['terminal.external.osxExec'],
+            'terminal.external.windowsExec': this.terminalExternalPreferences['terminal.external.windowsExec']
+        };
+    }
+
+    /**
+     * Display a quick pick for user to choose a target workspace in opened workspaces.
+     */
+    protected async selectTerminalExternalCwd(): Promise<URI | undefined> {
+        const roots = this.workspaceService.tryGetRoots();
+        return this.quickPickService.show(roots.map(
+            ({ resource }) => ({ label: this.labelProvider.getName(resource), description: this.labelProvider.getLongName(resource), value: resource })
```

Comment:
Can you break it up so its easier to read?

```suggestion
            ({ resource }) => ({ 
              label: this.labelProvider.getName(resource), 
              description: this.labelProvider.getLongName(resource),
              value: resource
            })
```